src/sas/qtgui/Perspectives/SizeDistribution/SizeDistributionThread.py

```diff
@@ -0,0 +1,83 @@
+import logging
+import sys
+
+from sas.qtgui.Perspectives.SizeDistribution.SizeDistributionUtils import (
+    MaxEntParameters,
+    MaxEntResult,
+)
+from sas.sascalc.data_util.calcthread import CalcThread
+from sas.sascalc.poresize.SizeDistribution import sizeDistribution
+from sasdata.dataloader.data_info import Data1D
+
+logger = logging.getLogger(__name__)
+
+
+class SizeDistributionThread(CalcThread):
+    """Thread performing the fit"""
+
+    def __init__(
+        self,
+        data: Data1D,
+        background: Data1D,
+        params: MaxEntParameters,
+        error_func=None,
+        completefn=None,
+        updatefn=None,
+        yieldtime=0.01,
+        worktime=0.01,
+    ):
+        """
+        Initialize parameters
+        """
+        CalcThread.__init__(self, completefn, updatefn, yieldtime, worktime)
+        self.data = data
+        self.background = background
+        self.params = params
+        self.error_func = error_func
+        self.starttime = 0
+
+    def compute(self, *args, **kwargs):
+        try:
+            sd = sizeDistribution(self.data)
+            sd.qMin = self.params.qmin
+            sd.qMax = self.params.qmax
+            sd.diamMin = self.params.dmin
+            sd.diamMax = self.params.dmax
+            sd.aspectRatio = self.params.aspect_ratio
+            sd.contrast = self.params.contrast
+            sd.model = self.params.model
+            sd.iterMax = self.params.max_iterations
+            sd.skyBackground = self.params.sky_background
+            sd.useWeights = True
+            sd.weightType = self.params.weight_type
+            sd.weightFactor = self.params.weight_factor
+            sd.nbins = self.params.num_bins
+
+            trim_data, intensities, init_bins_back, sigma = sd.prep_maxEnt(
+                self.background, full_fit=self.params.full_fit
+            )
+            convergence_info = sd.run_maxEnt(
+                trim_data, intensities, init_bins_back, sigma
+            )
+
+            convergences, num_iters = zip(*convergence_info)
+            results = MaxEntResult(
+                convergences=convergences,
+                num_iters=num_iters,
+                chisq=sd.chiSq_maxEnt,
+                bins=sd.bins,
+                bin_diff=sd._binDiff,
+                bin_mag=sd.BinMagnitude_maxEnt,
+                bin_err=sd.BinMagnitude_Errs,
+                data_max_ent=sd.Iq_maxEnt,
+                statistics=sd.MaxEnt_statistics,
+            )
+
+            self.completefn(results)
+        except KeyboardInterrupt:
+            pass
+        except Exception:
+            if self.error_func is not None:
+                self.error_func(
+                    "Error in size distribution fit: %s" % sys.exc_info()[1]
+                )
```

Comment:
Exceptions are already handled in the base CalcThread class in `CalcThread._run()`.

Reply:
We are using this to add custom error handling, like re-enabling the fitting buttons in the GUI. Ah but I see now that we can initialize CalcThread with this function as `exception_handler` instead. I will fix that.